docs/spec_draft.md

```diff
@@ -1267,6 +1268,70 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Applies a `body` function to `inputs` and `init_values` along the `dimensions`
+and produces a `result` tensor.
+
+The `body` function has to be associative, and it produces an
+implementation-defined behavior otherwise. Similarly, if `init_values` do not
+form identity under the `body` function, it produces an implementation-defined
+behavior.
+
+### Inputs
+
+| Name          | Type                                             |
+|---------------|--------------------------------------------------|
+| `inputs`      | variadic number of tensors of any supported type |
+| `init_values` | variadic number of scalars of any supported type |
+| `dimensions`  | 1-dimensional tensor constant of type `si64`     |
+| `body`        | `function`                                       |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `results` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) `inputs`, `init_values` have the same element type.
+  * (C2) All `inputs` have the same shape.
+  * (C3) type(`inputs[k]`) $=$ type(`init_values[k]`) for all `k` $\in$ [0, N).
+  * (C4) size(`inputs`) $=$ size(`init_values`) $=$ size(`result`) $=$ N where
+  N >= 1.
+  * (C5) `init_values` have to form an identity under the `body`.
+  * (C6) 0 $\lt$ `dimensions[d]` $\lt$ rank(`inputs[d]`) for all dimension `d`.
+  * (C7) All dimensions in `dimensions` are unique.
+  * (C8) The full function type of `body` is
```

Comment:
Even though I suggested the constraint C8, but it seems I was not entirely correct. Following up on https://github.com/openxla/stablehlo/issues/369, there is a bit of additional information that we might need to convey. 

How about we say 
`(T0,..., TN-1, T'0, ..., T'N-1) -> (T0, ..., TN-1)`, where
 * Tk corresponds to the element type of inputs[k] or `results[k]` ignoring floating-point precision.
 * T'k corresponds to the element type of `init_values[k]`    ignoring floating-point precision.

Reply:
`SameElementTypeIgnoringFpPrecision` seems to be a frequent occurrence in shape_inference.cc (e.g. it's used for concat, pad, binary, map, etc), but personally I've always treated it as a quirk of client HLO rather than something that we should follow in MHLO or StableHLO. What are the pros and cons of adding this feature to the spec? What is the motivation on the HLO side?

Reply:
>> What is the motivation on the HLO side?

This allows HLO verifier to accomodate mixed precision in HLO instruction, that is the operands, output, and parameters/output of the called computations can have different precisions (BF16 and F32).

There exists [passes](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/bfloat16_normalization.h) to add converters (f32->bf16 and reverse) so as to normalize the mixed precision version of the program.

If the implementation does not support bf16 or mixed precision, then the converters are folded away using [link](https://github.com/tensorflow/tensorflow/blob/b2ad6ba8c77da03fb10a894699a0c1a850a4322f/tensorflow/compiler/xla/service/bfloat16_conversion_folding.h#L25).

>> What are the pros and cons of adding this feature to the spec?

### Pros
The StableHLO spec will be aligned verbatim with the verifier. 

### Cons
To me it seems like a details tied to the implementation and would increase the complexity of the spec.
 
My take:
Even if we do not add this detail (which, more specifically, is a relaxation to type check), user can fully rely on the spec only to write  "type verified" program. After thinking more about it, I am aligned with your though process.

Reply:
"Given that MHLO doesn't support mixed precision". Actually I was wrong about this, and #369 goes into details of where exactly mixed precision is currently supported in MHLO (and in StableHLO).

Since this is a cross-cutting concern, i.e. it applies to a bunch of ops - not just `reduce` and not limited to the ops that #369 is talking about - I propose that we don't address it in this PR, but instead write the spec as if mixed precision is not a thing and then address all the ops at once.